charts/camunda-platform/test/integration/connection.go

```diff
@@ -0,0 +1,94 @@
+package integration
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/cookiejar"
+	"time"
+
+	"github.com/camunda-cloud/zeebe/clients/go/pkg/zbc"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/retry"
+	"google.golang.org/grpc"
+)
+
+func (s *integrationTest) createPortForwardedClient(serviceName string) (zbc.Client, func(), error) {
+	// NOTE: this only waits until the service is created, not until the underlying pods are ready to receive traffic
+	k8s.WaitUntilServiceAvailable(s.T(), s.kubeOptions, serviceName, 90, 1*time.Second)
+
+	// port forward the gateway service to avoid having to set up a public endpoint that the test can access externally
+	localGatewayPort := k8s.GetAvailablePort(s.T())
+	tunnel := k8s.NewTunnel(s.kubeOptions, k8s.ResourceTypeService, serviceName, localGatewayPort, 26500)
+
+	// the gateway is not ready/receiving traffic until at least one leader is present
+	s.waitUntilPortForwarded(tunnel, 30, 2*time.Second)
+
+	endpoint := fmt.Sprintf("localhost:%d", localGatewayPort)
+	client, err := zbc.NewClient(&zbc.ClientConfig{
+		GatewayAddress:         endpoint,
+		DialOpts:               []grpc.DialOption{},
+		UsePlaintextConnection: true,
+	})
+	if err != nil {
+		return nil, tunnel.Close, err
+	}
+
+	return client, func() { client.Close(); tunnel.Close() }, nil
+}
+
+func (s *integrationTest) createPortForwardedHttpClientWithPort(serviceName string, port int) (string, func()) {
+	return s.createPortForwardedHttpClientWithPortAndContainerPort(serviceName, port, 8080)
+}
+
+func (s *integrationTest) createPortForwardedHttpClientWithPortAndContainerPort(serviceName string, port int, containerPort int) (string, func()) {
+	// NOTE: this only waits until the service is created, not until the underlying pods are ready to receive traffic
+	k8s.WaitUntilServiceAvailable(s.T(), s.kubeOptions, serviceName, 90, 1*time.Second)
+
+	// remote port needs to be container port - not service port!
+	tunnel := k8s.NewTunnel(s.kubeOptions, k8s.ResourceTypeService, serviceName, port, containerPort)
+
+	// the gateway is not ready/receiving traffic until at least one leader is present
+	s.waitUntilPortForwarded(tunnel, 30, 2*time.Second)
```

Comment:
:question: I don't know if this is necessary for every HTTP port. I assume for most applications, having the port open is enough to show it's ready. That was just for Zeebe initially.

Reply:
It can happen that this fails. I would keep this.

Example:
```
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:33+02:00 retry.go:103: Waiting to port forward for endpoint localhost:8080 returned an error: Service camunda-platform-it-identity is not available. Sleeping for 2s and will try again.
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 retry.go:91: Waiting to port forward for endpoint localhost:8080
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 tunnel.go:153: Creating a port forwarding tunnel for resource svc/camunda-platform-it-identity routing local port 8080 to remote port 8080
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 client.go:42: Configuring Kubernetes client using config file /home/zell/.kube/config with context gke_zeebe-io_europe-west1-b_zeebe-cluster
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 client.go:42: Configuring Kubernetes client using config file /home/zell/.kube/config with context gke_zeebe-io_europe-west1-b_zeebe-cluster
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 client.go:42: Configuring Kubernetes client using config file /home/zell/.kube/config with context gke_zeebe-io_europe-west1-b_zeebe-cluster
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 tunnel.go:182: Error finding available pod: Service camunda-platform-it-identity is not available
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:35+02:00 retry.go:103: Waiting to port forward for endpoint localhost:8080 returned an error: Service camunda-platform-it-identity is not available. Sleeping for 2s and will try again.
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 retry.go:91: Waiting to port forward for endpoint localhost:8080
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 tunnel.go:153: Creating a port forwarding tunnel for resource svc/camunda-platform-it-identity routing local port 8080 to remote port 8080
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 client.go:42: Configuring Kubernetes client using config file /home/zell/.kube/config with context gke_zeebe-io_europe-west1-b_zeebe-cluster
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 client.go:42: Configuring Kubernetes client using config file /home/zell/.kube/config with context gke_zeebe-io_europe-west1-b_zeebe-cluster
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 client.go:42: Configuring Kubernetes client using config file /home/zell/.kube/config with context gke_zeebe-io_europe-west1-b_zeebe-cluster
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 tunnel.go:185: Selected pod camunda-platform-it-identity-8fbbb8ddd-44gps to open port forward to
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 tunnel.go:198: Using URL https://34.76.98.248/api/v1/namespaces/camunda-platform-s6nj3k/pods/camunda-platform-it-identity-8fbbb8ddd-44gps/portforward to create portforward
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 tunnel.go:247: Successfully created port forwarding tunnel
TestIntegration/TestServicesEnd2End 2022-04-27T11:52:37+02:00 connection.go:94: Endpoint localhost:8080 is now forwarded
```